### ISSUE TITLE
Use PreparedStatement.setObject(int, Object, Types.OTHER) instead of setString(int, String) to bind array bind values in PostgreSQL

### ISSUE THREAD

**Issue by mihalyr** [jOOQ/jOOQ#12485]
Hello,

I'm new to JOOQ and PostgreSQL, I was trying to move rows between tables with the following approach:

1. starting with `SELECT ID FROM QUEUED FOR UPDATE SKIP LOCKED LIMIT 2000`
1. then `INSERT INTO IN_FLIGHT SELECT FROM QUEUED q WHERE q.ID = ANY(selectedIds)`
1. and finally `DELETE FROM QUEUED WHERE ID = ANY(selectedIds)`

I've noticed that using ` = ANY(?::uuid[])` gets very slow with large sets compared to `IN (?, ?, ?)`.

Here are some results from local testing (PostgreSQL 11, JOOQ 3.13.3):

1. Using ` = ANY(?::uuid[])`

| iteration | batch size | time millis | batches for 2k | time for 2k |
|-----------|------------|-------------|----------------|-------------|
| 0         | 1          | 2.091       | 2048           | 4282.368    |
| 1         | 2          | 0.026       | 1024           | 26.624      |
| 2         | 4          | 0.024       | 512            | 12.288      |
| 3         | 8          | 0.038       | 256            | 9.728       |
| 4         | 16         | 0.092       | 128            | 11.776      |
| 5         | 32         | 0.115       | 64             | 7.36        |
| 6         | 64         | 0.217       | 32             | 6.944       |
| 7         | 128        | 0.377       | 16             | 6.032       |
| 8         | 256        | 0.753       | 8              | 6.024       |
| 9         | 512        | 962.491     | 4              | 3849.964    |
| 10        | 1024       | 1933.769    | 2              | 3867.538    |
| 11        | 2048       | 3740.711    | 1              | 3740.711    |

2. Using `IN (?, ?, ...)`

| iteration | batch size | time millis | batches for 2k | time for 2k |
|-----------|------------|-------------|----------------|-------------|
| 0         | 1          | 0.008       | 2048           | 16.384      |
| 1         | 2          | 0.026       | 1024           | 26.624      |
| 2         | 4          | 0.018       | 512            | 9.216       |
| 3         | 8          | 0.028       | 256            | 7.168       |
| 4         | 16         | 0.042       | 128            | 5.376       |
| 5         | 32         | 0.084       | 64             | 5.376       |
| 6         | 64         | 2.2         | 32             | 70.4        |
| 7         | 128        | 3.9         | 16             | 62.4        |
| 8         | 256        | 7.2         | 8              | 57.6        |
| 9         | 512        | 13.1        | 4              | 52.4        |
| 10        | 1024       | 23.08       | 2              | 46.16       |
| 11        | 2048       | 29.393      | 1              | 29.393      |

The times shown above are milliseconds as reported by `EXPLAIN ANALYZE`.

So, there is this significant jump in times around 512 elements when using `ANY` compared to `IN`, which still finished in double-digit milliseconds even with 2k items.

When debugging the code I noticed that when using `IN` JOOQ will delegate each UUID to the driver, which uses binary transfer for each of the UUIDs in the set. However, when using `ANY`, JOOQ will first convert the array to a String and pass that to the driver. Later I found out that the binary transfer is probably not relevant in this issue.

I was able to reproduce the same issue with the below JDBC snippet. The difference between the commented and uncommented lines is almost 4 seconds and JOOQ is using the slower `setString` and I cannot find a way to tell it to use the `setArray` method. What is even more interesting that `setArray` under the hood also converts the UUID array to String and calls the same protected `setString` as the other method, the only difference is that it sends the OID of the UUID array type.

```java
      tx.connection(conn -> {
        var prepared = conn.prepareStatement(
            "EXPLAIN (ANALYZE, COSTS, VERBOSE, BUFFERS, FORMAT JSON) " +
                "select id from mytable where id = any(?::uuid[])"
        );

        // prepared.setString(1, PostgresUtils.toPGArrayString(selectedIds));
        prepared.setArray(1, conn.createArrayOf("uuid", selectedIds));

        var resultSet = prepared.executeQuery();
        while (resultSet.next()) {
          System.out.println(resultSet.getString(1));
        }
      });
```

PS: I did some reading about my problem in general and I'm aware that this is not the ideal approach to move rows between tables, I'm considering switching to temporary tables later to keep the IDs in the database without sending them back and forth.

### Versions

- jOOQ: 3.13.3
- Java: 11 
- Database (include vendor): PostgreSQL 11
- OS: Fedora 34, kernel 5.13.19
- JDBC Driver (include name if inofficial driver):  42.2.14

**Comment by mihalyr** [comment]
I've tried one more thing

```java
        var arrString = PostgresUtils.toPGArrayString(selectedIds);
        var arr = new PgArray(conn.unwrap(PgConnection.class), Oid.UUID_ARRAY, arrString);
        prepared.setArray(1, arr);
```

This is also fast. So the problem does not seem to be with what we send and how it is serialized, but with how it is being passed to the driver, probably the OID affects something internally.

**Comment by lukaseder** [comment]
Thanks a lot for your investigation. These are curious findings, and while I'm not convinced that we should eagerly take action here in jOOQ, it's great to know that there are such significant differences in performance, depending on the bind type.

Binding arrays and UDTs in PostgreSQL is quite hard. For historic reasons, pgjdbc never implemented the JDBC `SQLData` API for UDTs (see e.g. https://github.com/pgjdbc/pgjdbc/issues/641). An entire alternative driver was implemented in parts because of this missing functionality: https://github.com/jOOQ/jOOQ/issues/10092#issuecomment-623197645. I don't recommend using it, as it appears to cause a lot of trouble elsewhere, but I do regret the competition rather than cooperation between the official and inofficial drivers.

I'm bringing this up, because any potential fix for your issue in jOOQ will open a can of worms in the area of binding nested UDTs containing arrays, containing UDTs, containing arrays, etc (which is the purpose of `PostgresUtils`, among other things). I'd rather not touch this area for a thing that can be worked around easily. All while any potential performance-only related workaround in jOOQ might be wrong/obsolete, as soon as some underlying issue in pgjdbc might be fixed (not sure if there is any, but it's not improbable).

So, there are a few workarounds:

- Use the `IN` predicate, as you've noticed (though, in my own benchmarks, the `= ANY(array)` syntax should outperform the `IN` predicate for large arrays: https://blog.jooq.org/sql-in-predicate-with-in-list-or-with-array-which-is-faster/, so this seems to be a weird `UUID[]` related edge case)
- Use inline values instead of bind values (have you tried this?)
- Re-organise your architecture to avoid `IN` predicates

To summarise:

- This seems to be an edge case
- It touches the very complex topic of binding nested types
- It can be worked around

Given the above, I prefer not to take any action at this point.

**Comment by mihalyr** [comment]
Thank you, Lukas, for taking a look so quickly.
It's insane how much complexity lives under the hood even when dealing with a innocent looking method call :slightly_smiling_face:  I really appreciate your detailed answer and I agree with your summary. For now this is a weird edge case that I can work around.

> Use inline values instead of bind values (have you tried this?)

I haven't, the problem is that I use this statement very frequently (and concurrently from many threads and servers) and my assumption was that inlining the values would require the calculation of new execution plans for every query. The other thing I noticed when using `IN` was that at least in GCP Cloud SQL insights the statements were not aggregated and it makes them less useful, which was one of my reasons to use `ANY(array)` in the first place and I thought that inline would lead to the same.

**Comment by lukaseder** [comment]
> and my assumption was that inlining the values would require the calculation of new execution plans for every query

Yes, it does, but this is PostgreSQL, which doesn't really have an execution plan cache, so it won't matter much.

> The other thing I noticed when using `IN` was that at least in GCP Cloud SQL insights the statements were not aggregated and it makes them less useful, which was one of my reasons to use `ANY(array)` in the first place and I thought that inline would lead to the same.

You could use `IN` list padding to work around that to some extent:
https://www.jooq.org/doc/latest/manual/sql-building/dsl-context/custom-settings/settings-in-list-padding/

Of course, it would again make your SQL slower on average...

**Comment by mihalyr** [comment]
> You could use IN list padding to work around that to some extent:

Very interesting, the amount of features in JOOQ keeps surprising me, thanks for the tip.

By the way I just tested `ID.eq(DSL.any(DSL.inline(ids)))` now and it also works as well as `IN` or the `setArray` method used above, so that is also a good alternative for my case.

> but this is PostgreSQL, which doesn't really have an execution plan cache, so it won't matter much.

I am new to Postgres and didn't know how the query plans are re-used (or not), maybe that explains what I saw in my tests. 
Below table shows "Planning time" from the explain analyze output and I don't see any improvement for the "not inlined" case there.

iteration | batch size | no inline | with inline
-- | -- | -- | --
0 | 1 | 0.096 | 0.102
1 | 2 | 0.098 | 0.087
2 | 4 | 0.114 | 0.097
3 | 8 | 0.163 | 0.104
4 | 16 | 0.12 | 0.089
5 | 32 | 0.146 | 0.102
6 | 64 | 0.196 | 0.146
7 | 128 | 0.251 | 0.111
8 | 256 | 0.497 | 0.143
9 | 512 | 0.698 | 0.179
10 | 1024 | 1.317 | 0.315
11 | 2048 | 2.485 | 0.757

Thanks again Lukas, I think I have all the info I need to fix this on my end.

**Comment by U1F984** [comment]
Hi everyone,

after a long debugging session I found this ticket. I have the same issue as the thread starter, with =ANY queries being slow (40 seconds instead of 5 milliseconds in our case for an array of ~20 000 elements).

Unfortunately following the advice of https://blog.jooq.org/sql-in-predicate-with-in-list-or-with-array-which-is-faster/ and similar blog posts & doing benchmarks we implemented the =ANY pattern all over our codebase. This had the additional benefit of making the query logs much easier to analyze since only a single type would be recognized.

Recently a query plan decided to switch from an index scan to a full table scan. Since the table is not much bigger than the specified 20 000 elements, this should be correct since most of the table is returned. With the index the lookup is fast since (presumably) Postgres iterates the array once, looking up each element, but with the sequential scan Postgres scans the array for each row. 

If the Parameter is passed via `setArray` Postgres can (again, presumably) apply an internal optimization to hash the array contents beforehand since the type is known.  With the current `setString` implementation in the `org.jooq.impl.DefaultBinding.DefaultArrayBinding#set0` however the query is slow, so despite the explicit cast in the query it cannot apply the optimization. If the query is manually prepared and executed (via SQL) it works as expected, so this might also be a limitation of the Postgres query protocol V3.

If the `set0` method is adjusted so it checks `value instanceof Long[]` and uses the `setArray` the query takes the expected time. 

Adding support for all primitive types + String in the `set0` would probably cover many usecases and provide a huge performance benefit for cases like the one described. Would you accept such a contribution?

**Comment by U1F984** [comment]
I was also wondering how this issue was not reproducible with `PREPARE/EXECUTE` when testing via psql. Turns out the issue is reproducible if the type is also given as `TEXT` in the initial `PREPARE`. That means an alternative is to set the type to `UNSPECIFIED` after the regular `setString` call:

```
((PgPreparedStatement) stmt.unwrap(PGStatement.class)).preparedParameters.getTypeOIDs()[index - 1] = Oid.UNSPECIFIED
```

That also gives a faster execution plan, but I think it would be the wrong solution for inclusion into jooq. Unfortunately the `org.postgresql.jdbc.PgPreparedStatement#setString(int, java.lang.String, int)` is not public...

**Comment by lukaseder** [comment]
> Adding support for all primitive types + String in the `set0` would probably cover many usecases and provide a huge performance benefit for cases like the one described. Would you accept such a contribution?

Let's discuss this further, first.

You could create a new issue, if you're certain this should be addressed from within jOOQ. I would need to see a reproducer with execution plans, though, giving a strong hint that one thing is truly better than the other. SQL performance advice tends to be very ad-hoc and volatile. Things that were slow yesterday no longer are, so I want to be sure this increased complexity actually improves things.

In fact, this very topic (using `ANY` instead of `IN`) just hints at performance issues in PostgreSQL or pgjdbc, which should be addressed there. The blog post just exposes this issue and suggests a *workaround*, not a *truly better* approach.

> That means an alternative is to set the type to `UNSPECIFIED` after the regular `setString` call:

We'll certainly not implement hacks like these, which may be a solution to your particular observation and case (if you were using JDBC directly), but will expose unknown side effects in many other cases.

How confident are you that this will improve things in general, not impair them? I wouldn't be, from that particular documentation. It looks like a hack for individual cases, not something to be used widely in a library like jOOQ.

Frankly, I don't understand this particular problem in PostgreSQL. If you look at the generated SQL of this query:

```java
ctx.fetchValue(val(new Integer[] { 1, 2 }));
```

jOOQ generates these things:

```sql
-- PreparedStatement
select cast(? as int[])

-- static Statement
select cast('{1,2}' as int[])
```

So, the bind value is clearly a string representation of the array (because that's the only sane way to encode more complex array types, e.g. using UDTs, nested arrays, range types, spatial types, etc. etc.) and it is cast explicitly to `int[]` by jOOQ, so I'm not sure why this produces a problem in PostgreSQL. Maybe, there's a bug there worth addressing? Maybe, the cast result (being a constant) is not cached in the query execution, and so it re-applied for every row, which would be a significant problem.

**Comment by U1F984** [comment]
Let me try add some more context. The issue can also be observed in this reproducer here (using pure SQL):

```sql
-- for copying to below
select '{' || array_to_string(array(select generate_series(1, 20000)), ',') || '}';

create temporary table demo(id bigint not null primary key, other_id bigint not null unique);
insert into demo (id, other_id)
select generate_series(1, 20000), generate_series(1, 20000);

prepare demo_ps(text) as select * from demo where other_id = any (cast($1 as bigint[]));

explain (analyze,verbose ,timing)
execute demo_ps('<the value from the first select>')
;
```
This will result in:
```
Seq Scan on pg_temp.demo  (cost=0.00..663.54 rows=20000 width=16) (actual time=2.611..47539.273 rows=20000 loops=1)
  Output: id, other_id
  Filter: (demo.other_id = ANY (('{1,2,3,4,5,6,<snipped>,20000}'::cstring)::bigint[]))
Query Identifier: 410052779080471777
Planning Time: 7.979 ms
Execution Time: 47543.906 ms
```

If one removes the explicit type declaration from the `PREPARE` call:

```sql
deallocate demo_ps;
prepare demo_ps as select * from demo where other_id = any (cast($1 as bigint[]));
```

The query is fast:

```
Seq Scan on pg_temp.demo  (cost=50.00..461.47 rows=20000 width=16) (actual time=0.788..3.928 rows=20000 loops=1)
  Output: id, other_id
  Filter: (demo.other_id = ANY ('{1,2,3,4,5,6,7,<snipped>,20000}'::bigint[]))
Query Identifier: -8997541469505792707
Planning Time: 6.897 ms
Execution Time: 7.636 ms
```

This is because postgres correctly inferred the type for the prepare statement as observed from `pg_prepared_statement`:

| name | statement | prepare\_time | parameter\_types |
| :--- | :--- | :--- | :--- |
| demo\_ps | prepare demo\_ps as select \* from demo where other\_id = any \(cast\($1 as bigint\[\]\)\) | 2024-08-05 12:18:01.712153 +00:00 | {bigint\[\]} |

This is also what is happenening with JDBC which can be seen in the pgjdbc trace logs:

```
# setting the stringtype to unspecified or using the hack (0 oid is unspecified type)
 org.postgresql.core.v3.QueryExecutorImpl sendParse  FE=> Parse(stmt=null,query="select * from "demo" where "other_id" = any (cast($1 as bigint[]))",oids={0})
# using setArray (1016 oid is bigint[])
 org.postgresql.core.v3.QueryExecutorImpl sendParse  FE=> Parse(stmt=null,query="select * from "demo" where "other_id" = any (cast($1 as bigint[]))",oids={1016})
# standard jooq behavior (1043 oid is character varying)
 org.postgresql.core.v3.QueryExecutorImpl sendParse  FE=> Parse(stmt=null,query="select * from "demo" where "other_id" = any (cast($1 as bigint[]))",oids={1043})
```

The first two are fast (inferred bigint[] and bigint[] specified explicitly), the last one is slow like with the pure SQL `PREPARE` (since the type is specified as TEXT).

One could argue that this is a postgres issue, since postgres should "see" that the text value passed in during `PREPARE` is cast to `bigint[]` and thus the optimization should be enabled. 

For those cases where the type is a known standard type such as `bigint[]`/`Long[]` I also think using the proper `setArray` calls in jOOQ would be an improvement in any case, even if that was fixed, since it enables other features such as binary transfer and avoids building the string. 

For the fallback case or if the special casing for standard types is undesirable, I also found the intended way to set the OID to unspecified using (pg)jdbc:

```java
// this is the existing code, sends oid = character varying
ctx.statement().setString(ctx.index(), toPGArrayString(value));
// this is the suggested improved code, sends oid = unspecified
ctx.statement().setObject(ctx.index(), toPGArrayString(value), Types.OTHER);
```

**Comment by lukaseder** [comment]
Thanks a lot for your detailed analysis, @U1F984. I can see your point.

> One could argue that this is a postgres issue, since postgres should "see" that the text value passed in during `PREPARE` is cast to `bigint[]` and thus the optimization should be enabled.

I would definitely bring this up on the PostgreSQL mailing lists. This is an optimiser limitation that I think they should address, or at least be aware of, for the time being.

I'll think about improving this.

**Comment by lukaseder** [comment]
Looking into this now.

I can live with this solution. It seems to make the least assumptions and doesn't produce any regressions in our integration tests.

> ```java
> // this is the suggested improved code, sends oid = unspecified
> ctx.statement().setObject(ctx.index(), toPGArrayString(value), Types.OTHER);
> ```

I rest my case that this is a very significant flaw to have in the PG optimiser, and it should be addressed there. I think this comment here could be forwarded to the PostgreSQL mailing lists (their "bug tracker"), if you haven't already, @U1F984:
https://github.com/jOOQ/jOOQ/issues/12485#issuecomment-2269003852

See mailing list here:
https://www.postgresql.org/list/pgsql-bugs/

**Comment by lukaseder** [comment]
Note that I couldn't reproduce the performance issue itself, with a few tests. Neither on PostgreSQL 12 nor 15, which I tried.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.20.0

To minimise regression risk (this affects *all* array bindings), I won't backport the fix, which isn't an "obvious" fix for the better. If you have some time to confirm the fix addresses your issues using 3.20.0-SNAPSHOT (from here if you're licensed: https://www.jooq.org/download/versions, or built from github otherwise), I'd greatly appreciate your feedback.

**Comment by lukaseder** [comment]
Thanks for this test case, @U1F984. I can reproduce this difference with or without the fix:

```
Total: 19.043s   (before the fix)
Total: 676.958ms (after the fix)
```

So, that confirms it for me.
